### DISCUSSION TITLE
self::getContainer()->get cant get Autowired Service

### DISCUSSION THREAD

**Discussion by alessandro-podo** [symfony/symfony#45242] [Q&A]
Symfony 5.4: works
$builder = self::getContainer()->get(EndPointBuilder::class);
Symfony 6.0.3: fail
$builder = self::getContainer()->get(EndPointBuilder::class);

Error:
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The "App\Service\CableLengthCalculator\Calculator" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

Did I miss something in the upgrade instructions?

**Comment by chyshyya** [comment]
I think you need to make your service public:
https://symfony.com/doc/current/service_container/alias_private.html

And some info about injecting services, but i think you know that.
Read news first (second part):
https://symfony.com/blog/new-in-symfony-5-4-controller-changes
And then read current docs (pass services to constructor or controller function as argument)
https://symfony.com/doc/current/service_container.html

**Comment by ahundiak** [comment]
@alessandro-podo Follow the instructions here: https://symfony.com/doc/current/testing.html#retrieving-services-in-the-test In particular you should be using static::getContainer instead of self::getContainer.  And you might need to add your service to config/services_test.yaml.

**Comment by alessandro-podo** [comment]
Thanks for the answers. 
After I set the services public, it works. 

But I just wonder why it changed with the update to 6.0.

**Comment by alessandro-podo** [comment]
After some trial and error, I found the solution. 
The TestContainer contains only the services that are loaded somewhere else via the constructor. Then you don't have to set them to public. 

is this with known behavior?

**Comment by ivanbogomoloff** [comment]
It's because all services private by default. See what is private services. You can't get private service from container like this

```php
$container->get(MyService::class); // <-- fail because it's private except if you describe this services as public in services.yaml
```

For test you need `services_test.yaml` which will configure and make your services accessable by `container->get(...)`

```yaml
services:
  _defaults:
    public: true # <-- this means that all services described in this file will be public by default and will be accessable by container->get

  App\Factory\UserFactory: ~ # <-- will be accessable
```

**Reply by vince83110** [reply]
Do you really need the service as public if you make it accessible ? Or the opposite by the way ? 

From doc:
`The container from static::getContainer() is actually a special test container. It gives you access to both the public services and the non-removed`
